### PR TITLE
Ghost qol: health scanner, various teleporter support

### PR DESCRIPTION
**Pull request by TheOneAndMightyRed**
## About The Pull Request

Ghosts get a health scanner, toggleable in the Ghost action tab at the top left.
grafik
Added support for ghost interaction (by clicking on said objects) for:
- Medevac beacon and stretcher
- Linked ASRS teleporters
- Xeno tunnels
- Wraith portals
- Shuttle computer

## Why It's Good For The Game

As it turns out, people die a lot in a PvP server and become ghosts for a good 30 minutes.
Being able to follow people through various teleportation methods as a ghost good.
Orbit menu searching and WASDing around the map to get to the linked teleporter/ portal bad.
Easy way for ghosts to go shipside/ groundside good.
Using convoluted jump menu in ghost tab bad.
Health scanner so people can point and laugh at the guy who got forcefed 4 inaprov pills and is unconcious for 20 minutes straight good.

## Changelog
:cl:
qol: Ghosts get a health scanner. Toggable in the ghost action tab at the top left.
qol: Ghosts can teleport through ASRS deployable teleporters.
qol: Ghosts can teleport through Medevac equipment.
qol: Ghosts can teleport through Xeno tunnels.
qol: Ghosts can teleport through Wraith portals.
qol: Ghosts can teleport using the Shuttle control computers located at each LZ and CIC.
/:cl:

### CODE REVIEW

**Comment by Wayland-Smithy** [comment]
Very nice! I've been meaning to hookup ghost attack for teleporters and portals. Mind adding `/obj/effect/wraith_portal` to your list? Just needs to check if `linked_portal` is valid and force move to it.

File: code/game/objects/structures/stool_bed_chair_nest/bed.dm [hunk 1]
```diff
@@ -364,6 +364,12 @@ GLOBAL_LIST_EMPTY(activated_medevac_stretchers)
 /obj/structure/bed/medevac_stretcher/attack_hand_alternate(mob/living/user)
 	activate_medevac_teleport(user)
 
+/obj/structure/bed/medevac_stretcher/attack_ghost(mob/dead/observer/user)
+	. = ..()
+	if(!linked_beacon || !linked_beacon.loc) //TODO: Redo medevac links
```
Comment (TheOneAndMightyRed) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:369]: This and the other medevac checks probably require a rewrite on how linking stretchers work in da future.
Outside the scope of PR therefore a TODO.

Comment (BraveMole) [inline code/game/objects/structures/stool_bed_chair_nest/bed.dm:369]: !linked_beacon?.loc should work. Same remark for the rest

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -40,6 +40,7 @@ GLOBAL_VAR_INIT(observer_default_invisibility, INVISIBILITY_OBSERVER)
 	var/inquisitive_ghost = FALSE
 	///A weakref to the original corpse of the observer
 	var/datum/weakref/can_reenter_corpse
+	var/obj/item/healthanalyzer/ghost/healthscanner = new()
```
Comment (BraveMole) [inline code/modules/mob/dead/observer/observer.dm:43]: No new in declaration , should be in Init

File: code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm [hunk 3]
```diff
@@ -517,6 +517,10 @@ GLOBAL_LIST_INIT(wraith_banish_very_short_duration_list, typecacheof(list(
 	QDEL_NULL(portal_visuals)
 	return ..()
 
+/obj/effect/wraith_portal/attack_ghost(mob/dead/observer/user)
+	. = ..()
+	linked_portal ? user.forceMove(get_turf(linked_portal)) : null
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm:522]: Use a if that is terrible

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/dead/observer/observer.dm [hunk 4]
```diff
@@ -40,6 +40,7 @@ GLOBAL_VAR_INIT(observer_default_invisibility, INVISIBILITY_OBSERVER)
 	var/inquisitive_ghost = FALSE
 	///A weakref to the original corpse of the observer
 	var/datum/weakref/can_reenter_corpse
+	var/obj/item/healthanalyzer/ghost/healthscanner
```
Comment (BraveMole) [inline code/modules/mob/dead/observer/observer.dm:43]: autodoc

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/game/objects/items/scanners.dm [hunk 5]
```diff
@@ -94,12 +99,18 @@ REAGENT SCANNER
...
-	update_static_data(user)
+	scan_target(patient)
 	if(user.skills.getRating("medical") < upper_skill_threshold)
 		return
 	START_PROCESSING(SSobj, src)
 
+/obj/item/healthanalyzer/proc/scan_target(mob/living/carbon/Target)
+	if(!iscarbon(Target) || isxeno(Target))
+		return
+	patient = Target
```
[the remark this answers is not in the document]
Comment (TheOneAndMightyRed) [reply code/game/objects/items/scanners.dm:110]: Need an example on how to do that. Can ya point me to a part of the code where that was implemented in?

**Comment by TheOneAndMightyRed** [comment]
Gonna reopen when i got time to fix stuff.

**Comment by TheOneAndMightyRed** [comment]
Hopefully git actually does what i want it to do, gimme a moment

**Comment by TheOneAndMightyRed** [comment]
wew it still works

**Comment by XSlayer300** [comment]
Will the refresh button ever work? Just reminding that since the topic of ghost stuffs.
image

**Comment by TheOneAndMightyRed** [comment]
Uuuh i assume that is on the orbit menu? I very rarely use the orbit menu so i don't know much about it. But lemme take a look anyway.

**Review by BraveMole — CHANGES_REQUESTED** [review]
I'm not sure I was clear enough on discord so here it is:

The idea is really good, but giving ghost an /obj/item/healanalyzer is not acceptable. The probably only good way of adding this feature is by refactoring health analyzer to be a component.

**Pull request closed without merging**